### PR TITLE
fix(postgres): use schema set in sequelize config by default

### PR DESCRIPTION
**Pull request by nholmes3**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This is an attempt at fixing a bug where the Postgres Query Generator does not respect a custom schema name if it is provided.

[This comment](https://github.com/sequelize/cli/issues/284#issuecomment-573351082) on [Issue #248 of sequelize/cli](https://github.com/sequelize/cli/issues/284) describes the problem in detail.

### CODE REVIEW

**Comment by nholmes3** [comment]
> Hi @nholmes3, thanks for the PR! Could you add some tests for this? Where you test the normal 'public' schema but also a custom schema

Hey @WikiRik I would be happy to try to add some tests. Do you mind pointing me at the file where they would be most appropriate? Thanks

File: test/unit/dialects/postgres/query-generator.test.js [hunk 1]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
...
+
+      const customSchemaSuites = {
+        showTablesQuery: [
+          {
+            title: 'showTablesQuery should use correct custom schema',
+            arguments: [],
+            expectation: 'SELECT table_name FROM information_schema.tables WHERE table_schema = \'custom\' AND table_type LIKE \'%TABLE\' AND table_name != \'spatial_ref_sys\';',
+          },
+        ],
+        describeTableQuery: [
```
Comment (nholmes3) [inline test/unit/dialects/postgres/query-generator.test.js:1357]: @ephys @WikiRik 

I added these unit tests to cover the postgres dialect. However, I am having some serious trouble getting the assertion for `describeTableQuery` to work. The strings that chai provides in the diff appear to be exactly equal. I am not sure what else to try to get this test to pass

AssertionError: expected 'SELECT pk.constraint_type as "Constra…' to equal 'SELECT pk.constraint_type as "Constra…'
      + expected - actual

-SELECT pk.constraint_type as "Constraint",c.column_name as "Field", c.column_default as "Default",c.is_nullable as "Null", (CASE WHEN c.udt_name = 'hstore' THEN c.udt_name ELSE c.data_type END) || (CASE WHEN c.character_maximum_length IS NOT NULL THEN '(' || c.character_maximum_length || ')' ELSE '' END) as "Type", (SELECT array_agg(e.enumlabel) FROM pg_catalog.pg_type t JOIN pg_catalog.pg_enum e ON t.oid=e.enumtypid WHERE t.typname=c.udt_name) AS "special", (SELECT pgd.description FROM pg_catalog.pg_statio_all_tables AS st INNER JOIN pg_catalog.pg_description pgd on (pgd.objoid=st.relid) WHERE c.ordinal_position=pgd.objsubid AND c.table_name=st.relname) AS "Comment" FROM information_schema.columns c LEFT JOIN (SELECT tc.table_schema, tc.table_name, cu.column_name, tc.constraint_type FROM information_schema.TABLE_CONSTRAINTS tc JOIN information_schema.KEY_COLUMN_USAGE  cu ON tc.table_schema=cu.table_schema and tc.table_name=cu.table_name and tc.constraint_name=cu.constraint_name and tc.constraint_type='PRIMARY KEY') pk ON pk.table_schema=c.table_schema AND pk.table_name=c.table_name AND pk.column_name=c.column_name WHERE c.table_name = 'myTable' AND c.table_schema = 'custom' 

+SELECT pk.constraint_type as "Constraint",c.column_name as "Field", c.column_default as "Default",c.is_nullable as "Null", (CASE WHEN c.udt_name = 'hstore' THEN c.udt_name ELSE c.data_type END) || (CASE WHEN c.character_maximum_length IS NOT NULL THEN '(' || c.character_maximum_length || ')' ELSE '' END) as "Type", (SELECT array_agg(e.enumlabel) FROM pg_catalog.pg_type t JOIN pg_catalog.pg_enum e ON t.oid=e.enumtypid WHERE t.typname=c.udt_name) AS "special", (SELECT pgd.description FROM pg_catalog.pg_statio_all_tables AS st INNER JOIN pg_catalog.pg_description pgd on (pgd.objoid=st.relid) WHERE c.ordinal_position=pgd.objsubid AND c.table_name=st.relname) AS "Comment" FROM information_schema.columns c LEFT JOIN (SELECT tc.table_schema, tc.table_name, cu.column_name, tc.constraint_type FROM information_schema.TABLE_CONSTRAINTS tc JOIN information_schema.KEY_COLUMN_USAGE  cu ON tc.table_schema=cu.table_schema and tc.table_name=cu.table_name and tc.constraint_name=cu.constraint_name and tc.constraint_type='PRIMARY KEY') pk ON pk.table_schema=c.table_schema AND pk.table_name=c.table_name AND pk.column_name=c.column_name WHERE c.table_name = 'myTable' AND c.table_schema = 'custom'

**Comment by nholmes3** [comment]
> This change does not exclusively impact postgres. I think these files would be better places to add the tests to: https://github.com/sequelize/sequelize/tree/main/test/unit/sql.
> 
> I would say add at test in:
> 
> * [add-column.test.js](https://github.com/sequelize/sequelize/blob/main/test/unit/sql/add-column.test.js)
> * [remove-column.test.js](https://github.com/sequelize/sequelize/blob/main/test/unit/sql/remove-column.test.js)
> 
> The test should use a custom sequelize instance which sets a default schema, and ensure the generated SQL uses that schema.
> 
> This test also impacts describeTable, so a similar test should be added in https://github.com/sequelize/sequelize/blob/main/test/integration/query-interface/describeTable.test.js

@ephys  I have added tests to add-column and remove-column, but I am a little bit overwhelmed by how to properly add a test to the `describeTable` integration tests. Do you mind providing some additional guidance?

File: test/unit/dialects/postgres/query-generator.test.js (hunk 1 above)
Comment (WikiRik) [reply test/unit/dialects/postgres/query-generator.test.js:1357]: There's an extra space at the end in the actual one apparently

Comment (WikiRik) [reply test/unit/dialects/postgres/query-generator.test.js:1357]: See https://github.com/sequelize/sequelize/pull/14634/files#diff-d6c44e5861b8b76b4562220dbed9b2043de963db1178b62507aaee6bc0dfdfdbL163

Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js:1357]: Ack! Thanks for the second set of eyes

File: test/unit/dialects/postgres/query-generator.test.js [hunk 2]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
         });
       });
     });
+
+    describe('Custom Schema Functions', () => {
+      beforeEach(function () {
+        this.queryGenerator = new QueryGenerator({
```
Comment (ephys) [inline test/unit/dialects/postgres/query-generator.test.js:1337]: create a new sequelize instance instead, with the option set (see other comments on how to create a new test instance)

File: test/unit/dialects/postgres/query-generator.test.js [hunk 3]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
...
+      });
+      afterEach(function () {
+        // Remove custom schema in order to not interfere with other tests
+        this.queryGenerator.sequelize.options.schema = null;
+      });
+
+      const customSchemaSuites = {
+        showTablesQuery: [
+          {
+            title: 'showTablesQuery should use correct custom schema',
```
Comment (ephys) [inline test/unit/dialects/postgres/query-generator.test.js:1352]:
```suggestion
            title: 'showTablesQuery defaults to the schema set in Sequelize options',
```

Same for the other 2

File: test/unit/dialects/postgres/query-generator.test.js [hunk 4]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
...
+            arguments: [],
+            expectation: 'SELECT table_name FROM information_schema.tables WHERE table_schema = \'custom\' AND table_type LIKE \'%TABLE\' AND table_name != \'spatial_ref_sys\';',
```
Comment (ephys) [inline test/unit/dialects/postgres/query-generator.test.js:1354]: Use a template string here to remove the need to escape `'`

File: test/unit/dialects/postgres/query-generator.test.js [hunk 5]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
         });
       });
     });
+
+    describe('Custom Schema Functions', () => {
```
Comment (ephys) [inline test/unit/dialects/postgres/query-generator.test.js:1335]:
```suggestion
    describe('With custom schema in Sequelize options', () => {
```

File: test/unit/dialects/postgres/query-generator.test.js [hunk 6]
```diff
@@ -1331,5 +1331,53 @@ if (dialect.startsWith('postgres')) {
...
+          },
+        ],
+        addColumnQuery: [
```
Comment (ephys) [inline test/unit/dialects/postgres/query-generator.test.js:1364]: this test can be removed, it is redundant with the one added in `add-column.test.js`

**Review by ephys — CHANGES_REQUESTED** [review]
The `describeTable` integration test can be skipped because we already have an integration test that checks the query that includes schemas is valid, so the unit test you added is sufficient :)

File: test/unit/sql/add-column.test.js [hunk 7]
```diff
@@ -2,11 +2,16 @@
...
 const Support   = require('../support');
 const { DataTypes } = require('@sequelize/core');
+const _ = require('lodash');
 
 const expectsql = Support.expectsql;
 const current   = Support.sequelize;
 const sql       = current.dialect.queryGenerator;
 
+const custom = _.cloneDeep(current);
+custom.options.schema = 'custom';
```
[the remark this answers is not in the document]
Comment (nholmes3) [reply test/unit/sql/add-column.test.js]: fixed

File: test/unit/sql/add-column.test.js [hunk 8]
```diff
@@ -66,3 +71,32 @@ if (['mysql', 'mariadb'].includes(current.dialect.name)) {
     });
   });
 }
+
+describe(`Custom Schema ${Support.getTestDialectTeaser('SQL')}`, () => {
+  describe('addColumnCustomSchema', () => {
+
+    const Model = current.define('users', {
```
[the remark this answers is not in the document]
Comment (nholmes3) [reply test/unit/sql/add-column.test.js]: fixed

File: test/unit/sql/add-column.test.js [hunk 9]
```diff
@@ -66,3 +71,32 @@ if (['mysql', 'mariadb'].includes(current.dialect.name)) {
...
+        allowNull: false,
+      })), {
+        mariadb: 'ALTER TABLE `users` ADD `level_id` FLOAT NOT NULL;',
+        mysql: 'ALTER TABLE `users` ADD `level_id` FLOAT NOT NULL;',
+        postgres: 'ALTER TABLE "custom"."users" ADD COLUMN "level_id" FLOAT NOT NULL;',
+        sqlite: 'ALTER TABLE `users` ADD `level_id` FLOAT NOT NULL;',
+        mssql: 'ALTER TABLE [users] ADD [level_id] FLOAT NOT NULL;',
+        db2: 'ALTER TABLE "users" ADD "level_id" FLOAT NOT NULL;',
+        snowflake: 'ALTER TABLE "users" ADD "level_id" FLOAT NOT NULL;',
+        ibmi: 'ALTER TABLE "users" ADD "level_id" FLOAT NOT NULL',
```
[the remark this answers is not in the document]
Comment (nholmes3) [reply test/unit/sql/add-column.test.js]: my understanding is that not every dialect even has the concept of a schema...These tests should just verify that the changes made for postgres do not break the other dialects

File: test/unit/sql/add-column.test.js [hunk 10]
```diff
@@ -66,3 +71,32 @@ if (['mysql', 'mariadb'].includes(current.dialect.name)) {
...
+
+    const Model = current.define('users', {
+      id: {
+        type: DataTypes.INTEGER,
+        primaryKey: true,
+        autoIncrement: true,
+      },
+    }, { timestamps: false });
+
+    it('properly generate alter queries', () => {
```
[the remark this answers is not in the document]
Comment (nholmes3) [reply test/unit/sql/add-column.test.js]: fixed

File: test/unit/sql/remove-column.test.js [hunk 11]
```diff
@@ -28,3 +33,22 @@ if (current.dialect.name !== 'sqlite') {
     });
   });
 }
+
+describe(`Custom Schema ${Support.getTestDialectTeaser('SQL')}`, () => {
```
[the remark this answers is not in the document]
Comment (nholmes3) [reply test/unit/sql/remove-column.test.js]: fixed

File: test/unit/dialects/postgres/query-generator.test.js (hunk 2 above)
Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js:1342]: fixed

File: test/unit/dialects/postgres/query-generator.test.js (hunk 3 above)
Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js]: fixed

File: test/unit/dialects/postgres/query-generator.test.js (hunk 4 above)
Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js]: fixed

File: test/unit/dialects/postgres/query-generator.test.js (hunk 6 above)
Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js]: removed

File: test/unit/dialects/postgres/query-generator.test.js (hunk 5 above)
Comment (nholmes3) [reply test/unit/dialects/postgres/query-generator.test.js]: fixed

**Review by ephys — APPROVED** [review]
This looks good to me.

@WikiRik if it's ok for you too, we can merge

**Review by WikiRik — APPROVED** [review]

**Comment by ephys** [comment]
Here we go :)
Thank you for your work on this!

**Comment by nholmes3** [comment]
@WikiRik @ephys Thank you for the help! Do you have any information on when I can expect these changes to be officially released? Thanks!

**Comment by nholmes3** [comment]
> We can make a new alpha release of v7 if you want, but if you want this in v6 we would like to ask you to make a new PR for that.

@WikiRik I would prefer to get this in v6. I checked the documentation for contributing and I don't see any explicit instructions for how to create a pull request for a release. Can you please provide additional guidance?

**Pull request merged**